### PR TITLE
[WIP][Prototype] Experimental repartition_by_hash using dynamic task-generation

### PR DESCRIPTION
**Pull request by rjzamora**
Requires [distributed#3765](https://github.com/dask/distributed/pull/3765)
Based on [dask#6173](https://github.com/dask/dask/pull/6173)

Warning!! This is currently very rough.

Usage:
```python
ddf.repartition(
    columns=[...],
    ignore_index=True,
    dynamic_tasks=True,
)
```

cc @madsbk

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/dask_cudf/dask_cudf/sorting.py [hunk 1]
```diff
@@ -18,9 +18,93 @@
...
+        self.stages = stages
+        self.inputs = inputs  # Input IDs
+        self.token = token
+        self.columns = columns
+        self.ignore_index = ignore_index
+        self.output_keys = []
+
+    def __repr__(self):
+        return (
+            f"All2All(input_keys={self.input_keys}, "
```
Comment (beckernick) [inline python/dask_cudf/dask_cudf/sorting.py:38]: @rjzamora can we remove this reference to `self.input_keys` unless we need the keys themselves as part of the repr? It currently doesn't exist in the class

File: python/dask_cudf/dask_cudf/sorting.py [hunk 2]
```diff
@@ -18,9 +18,87 @@
...
+                )
+                priorities[out_key_c] = priority
+                deps[out_key_c] = keys_c
+                tasks[out_key_c] = (_concat, keys_c, self.ignore_index)
+
+                out_key = str((shuffle_token, i))
+                in_key_c = str((shuffle_combine_token, self.stages, inp))
+                deps[out_key] = (in_key_c,)
+                tasks[out_key] = in_key_c
+                self.output_keys.append(out_key)
```
Comment (madsbk) [inline python/dask_cudf/dask_cudf/sorting.py:99]:
```suggestion                
        for i, inp in enumerate(self.inputs):
            out_key = str((shuffle_token, i))
            in_key_c = str((shuffle_combine_token, self.stages, inp))
            deps[out_key] = (in_key_c,)
            tasks[out_key] = in_key_c
            self.output_keys.append(out_key)                
```

**Comment by harrism** [comment]
@rjzamora is this still needed?

**Comment by rjzamora** [comment]
> @rjzamora is this still needed?

We should be able to close this, but I may need to keep the branch alive until Mads has a long-term solution in place.  @beckernick @madsbk - Do we need to keep this branch alive?

**Pull request closed without merging**